src/components/project/handlers/set-department-id.handler.ts

```diff
@@ -0,0 +1,85 @@
+import { ServerException } from '../../../common';
+import {
+  DatabaseService,
+  EventsHandler,
+  IEventHandler,
+  ILogger,
+  Logger,
+} from '../../../core';
+import { Project, ProjectStep } from '../dto';
+import { ProjectUpdatedEvent } from '../events';
+
+type SubscribedEvent = ProjectUpdatedEvent;
+
+@EventsHandler(ProjectUpdatedEvent)
+export class SetInitialEndDate implements IEventHandler<SubscribedEvent> {
+  constructor(
+    private readonly db: DatabaseService,
+    @Logger('project:set-department-id') private readonly logger: ILogger
+  ) {}
+
+  async handle(event: SubscribedEvent) {
+    this.logger.debug('Project mutation, set department id', {
+      ...event,
+      event: event.constructor.name,
+    });
+
+    const shouldSetDepartmentId =
+      event.updates.step === ProjectStep.PendingFinanceConfirmation &&
+      !event.updated.departmentId.value;
+    if (!shouldSetDepartmentId) {
+      return;
+    }
+
+    try {
+      await this.assignDepartmentIdForProject(event.updated);
+    } catch (exception) {
+      throw new ServerException(
+        'Could not set departmentId on project',
+        exception
+      );
+    }
+  }
+
+  private async getDepartmentIdPrefixForProject(project: Project) {
+    // TODO: validate if active property lives on relationship or BaseNode
+    const accountNumbers = await this.db
+      .query()
+      .raw(
+        `
+        MATCH (:Project {id: $projectId})-[:primaryLocation {active: true}]-()-[:fundingAccount {active: true}]-()-[:accountNumber {active:true}]-(accountNumberNode:Property)
+        RETURN accountNumberNode.value
+        `,
+        { projectId: project.id }
+      )
+      .run();
+    return accountNumbers[0]['accountNumberNode.value'];
```

Comment:
We should probably check if a result wasn't found and throw an exception here.
`.first()` could also be used as a shortcut (instead of `.run()`.